### PR TITLE
Change timer ack manager to have readlevel per cluster

### PR DESCRIPTION
**Pull request by vancexu**

### CODE REVIEW

File: service/history/shardContext.go [hunk 1]
```diff
@@ -742,22 +747,29 @@ func (s *shardContextImpl) emitShardInfoMetricsLogsLocked() {
...
+			cluster = clusterMetadata.ClusterNameForFailoverVersion(task.GetVersion())
+		}
+
+		if ts.Before(s.timerMaxReadLevelMap[cluster]) {
 			// This can happen if shard move and new host have a time SKU, or there is db write delay.
 			// We generate a new timer ID using timerMaxReadLevel.
-			s.logger.Warnf("%v: New timer generated is less than read level. timestamp: %v, timerMaxReadLevel: %v",
-				time.Now(), ts, s.timerMaxReadLevel)
-			persistence.SetVisibilityTSFrom(task, s.timerMaxReadLevel.Add(time.Millisecond))
+			s.logger.WithField("Cluster", cluster).Warnf("%v: New timer generated is less than read level. timestamp: %v, timerMaxReadLevel: %v,",
```
Comment (dmetzgar) [inline service/history/shardContext.go]: Why the extra comma at the end?

Comment (vancexu) [reply service/history/shardContext.go]: Thanks for catching, typo fixed.

File: service/history/timerQueueActiveProcessor.go [hunk 2]
```diff
@@ -74,6 +74,7 @@ func newTimerQueueActiveProcessor(shard ShardContext, historyService *historyEng
 		timeNow,
 		updateShardAckLevel,
 		logger,
+		currentClusterName,
```
Comment (wxing1292) [inline service/history/timerQueueActiveProcessor.go]: do you actually use this?

File: service/history/timerQueueStandbyProcessor.go [hunk 3]
```diff
@@ -71,6 +71,7 @@ func newTimerQueueStandbyProcessor(shard ShardContext, historyService *historyEn
 		timeNow,
 		updateShardAckLevel,
 		logger,
+		clusterName,
```
Comment (wxing1292) [inline service/history/timerQueueStandbyProcessor.go]: do you actually use this?

File: service/history/timerQueueStandbyProcessor_test.go [hunk 4]
```diff
@@ -127,6 +127,7 @@ func (s *timerQueueStandbyProcessorSuite) SetupTest() {
 		logger:                    s.logger,
 		domainCache:               cache.NewDomainCache(s.mockMetadataMgr, s.mockClusterMetadata, metricsClient, s.logger),
 		metricsClient:             metrics.NewClient(tally.NoopScope, metrics.History),
+		timerMaxReadLevelMap:      make(map[string]time.Time),
```
Comment (wxing1292) [inline service/history/timerQueueStandbyProcessor_test.go]: do you actually use this?

File: service/history/timerQueueAckMgr_test.go [hunk 5]
```diff
@@ -580,6 +583,7 @@ func (s *timerQueueFailoverAckMgrSuite) SetupTest() {
 		logger:                    s.logger,
 		domainCache:               cache.NewDomainCache(s.mockMetadataMgr, s.mockClusterMetadata, s.metricsClient, s.logger),
 		metricsClient:             s.metricsClient,
+		timerMaxReadLevelMap:      make(map[string]time.Time),
```
Comment (wxing1292) [inline service/history/timerQueueAckMgr_test.go]: do you actually use this?

File: service/history/timerQueueAckMgr_test.go [hunk 6]
```diff
@@ -134,6 +134,7 @@ func (s *timerQueueAckMgrSuite) SetupTest() {
...
```
Comment (wxing1292) [inline service/history/timerQueueAckMgr_test.go]: do you actually use this?

File: service/history/transferQueueActiveProcessor_test.go [hunk 7]
```diff
@@ -136,6 +137,7 @@ func (s *transferQueueActiveProcessorSuite) SetupTest() {
 		logger:                    s.logger,
 		domainCache:               cache.NewDomainCache(s.mockMetadataMgr, s.mockClusterMetadata, metricsClient, s.logger),
 		metricsClient:             metrics.NewClient(tally.NoopScope, metrics.History),
+		timerMaxReadLevelMap:      make(map[string]time.Time),
```
Comment (wxing1292) [inline service/history/transferQueueActiveProcessor_test.go]: same here

File: service/history/transferQueueStandbyProcessor_test.go [hunk 8]
```diff
@@ -130,6 +130,7 @@ func (s *transferQueueStandbyProcessorSuite) SetupTest() {
 		domainCache:               cache.NewDomainCache(s.mockMetadataMgr, s.mockClusterMetadata, metricsClient, s.logger),
 		metricsClient:             metrics.NewClient(tally.NoopScope, metrics.History),
 		standbyClusterCurrentTime: make(map[string]time.Time),
+		timerMaxReadLevelMap:      make(map[string]time.Time),
```
Comment (wxing1292) [inline service/history/transferQueueStandbyProcessor_test.go]: same here

File: service/history/shardContext.go [hunk 9]
```diff
@@ -311,19 +311,24 @@ func (s *shardContextImpl) UpdateDomainNotificationVersion(domainNotificationVer
...
 
-func (s *shardContextImpl) UpdateTimerMaxReadLevel() time.Time {
+func (s *shardContextImpl) UpdateTimerMaxReadLevel(cluster string) time.Time {
 	s.Lock()
 	defer s.Unlock()
 
-	s.timerMaxReadLevel = s.GetTimeSource().Now().Add(s.config.TimerProcessorMaxTimeShift())
-	return s.timerMaxReadLevel
+	currentTime := s.GetTimeSource().Now()
+	if cluster != "" && cluster != s.GetService().GetClusterMetadata().GetCurrentClusterName() {
```
[the remark this answers is not in the document]
Comment (vancexu) [reply service/history/shardContext.go]: For some test cases like `TestTimerQueueFailoverAckMgrSuite/TestReadCompleteUpdateTimerTasks`

File: service/history/timerQueueActiveProcessor.go (hunk 2 above)
Comment (vancexu) [reply service/history/timerQueueActiveProcessor.go]: Yes I need them otherwise code cannot run

**Pull request merged**